pkg/cloudprovider/provider/vsphere/helper.go

```diff
@@ -218,13 +211,80 @@ func createClonedVM(ctx context.Context, vmName string, config *Config, session
 	return virtualMachine, nil
 }
 
-func uploadAndAttachISO(ctx context.Context, session *Session, vmRef *object.VirtualMachine, localIsoFilePath, datastoreName string) error {
-	datastore, err := session.Finder.Datastore(ctx, datastoreName)
-	if err != nil {
-		return fmt.Errorf("failed to get datastore: %v", err)
+func resolveDatastoreRef(ctx context.Context, config *Config, session *Session, vm *object.VirtualMachine, folder *object.Folder, cloneSpec *types.VirtualMachineCloneSpec) (*types.ManagedObjectReference, error) {
+	// Based on https://github.com/vmware/govmomi/blob/v0.22.1/govc/vm/clone.go#L358
+	if config.DatastoreCluster != "" && config.Datastore == "" {
```

Comment:
I don't think you need to keep checking whether the value of `DatastoreCluster` and the `Datastore` repeatedly. You do validation on the machine creation and I think that's fairly enough.

Reply:
IMHO this is simple enough to stay as it is, I don't see any substantial gain in doing something like:

```
if config.DatastoreCluster != nil {
   ...
   return ...
}
if config.Datastore != nil {
   ...
   return ...
}
```
I agree that the validation should guarantee that just one of the two is present.

Reply:
I am not saying to do it as you suggested, you should do it as: 
```
if config.DatastoreCluster != "" {
 .....
} else {
// Since its gonna be either Datastore Cluster or Datastore, then this is a Datastore and the validation method can guarantee that!!
}
```

Please adjust accordingly as we should never cover something that will never happen!?

Reply:
Again, I tend to think that this is kind of cosmetic change. Your proposal is less redundant, mine is more defensive.

Reply:
Yours is simply checking for something that the added workflow doesn't satisfy. The `config.Datastore == ""` is always true  if the `config.DatastoreCluster != ""` and same for ` config.DatastoreCluster == ""` in the scond condition, it is always true if `config.Datastore != ""`. thus those checks are useless not defensive.

Reply:
If the validation makes his job indeed there is no reason why both are populated. I try to explain you what I meant with defensive. Consider those two scenarios:

1. The mutatingwebhookconfigurations for machine-controller gets deleted for some reason
2. In a further development we break accidentally the logic of the validation

Maybe it's a bit paranoid :-) but it was just to explain myself

Reply:
An implicit additional check costs nothing and adds more readability to what is expected than implicitly relying on the condition being checked by the validation code.